code/modules/research/rdconsole.dm

```diff
@@ -556,32 +553,70 @@ doesn't have toxins access.
 
 /obj/machinery/computer/rdconsole/proc/ui_techweb()		//Legacy code.
 	var/list/l = list()
-	var/list/avail = list()			//This could probably be optimized a bit later.
-	var/list/unavail = list()
-	var/list/res = list()
+	var/list/backlog = list()
 	for(var/v in stored_research.researched_nodes)
-		res += stored_research.researched_nodes[v]
-	for(var/v in stored_research.available_nodes)
-		if(stored_research.researched_nodes[v])
-			continue
-		avail += stored_research.available_nodes[v]
-	for(var/v in stored_research.visible_nodes)
-		if(stored_research.available_nodes[v])
-			continue
-		unavail += stored_research.visible_nodes[v]
-	l += "<h2>Technology Nodes:</h2>[RDSCREEN_NOBREAK]"
-	l += "<div><h3>Available for Research:</h3>"
-	for(var/datum/techweb_node/N in avail)
-		var/not_unlocked = (stored_research.available_nodes[N.id] && !stored_research.researched_nodes[N.id])
-		var/has_points = (stored_research.research_points >= N.get_price(stored_research))
-		var/research_href = not_unlocked? (has_points? "<A href='?src=[REF(src)];research_node=[N.id]'>Research</A>" : "<span class='linkOff bad'>Not Enough Points</span>") : null
-		l += "<A href='?src=[REF(src)];view_node=[N.id];back_screen=[screen]'>[N.display_name]</A>[research_href]"
-	l += "</div><div><h3>Locked Nodes:</h3>"
-	for(var/datum/techweb_node/N in unavail)
-		l += "<A href='?src=[REF(src)];view_node=[N.id];back_screen=[screen]'>[N.display_name]</A>"
-	l += "</div><div><h3>Researched Nodes:</h3>"
-	for(var/datum/techweb_node/N in res)
-		l += "<A href='?src=[REF(src)];view_node=[N.id];back_screen=[screen]'>[N.display_name]</A>"
+		backlog += stored_research.researched_nodes[v]
+
+	var/list/columns = list()
+	var/max_tier = 0
+	for (var/node_ in stored_research.tiers)
+		var/datum/techweb_node/node = node_
+		var/tier = stored_research.tiers[node]
+		LAZYINITLIST(columns["[tier]"])  // String hackery to make the numbers associative
+		columns["[tier]"] += ui_techweb_single_node(node, minimal=(tier != 1))
+		max_tier = max(max_tier, tier)
+
+	l += "<table><tr><th align='left'>Researched</th><th align='left'>Available</th><th align='left'>Future</th></tr><tr>[RDSCREEN_NOBREAK]"
+	for(var/tier in 0 to max_tier)
+		l += "<td valign='top'>[RDSCREEN_NOBREAK]"
+		l += columns["[tier]"]
+		l += "</td>[RDSCREEN_NOBREAK]"
+	l += "</tr></table>[RDSCREEN_NOBREAK]"
+	return l
+
+/obj/machinery/computer/rdconsole/proc/build_path_icon(atom/item, user)

```

Comment:
The reason they were saying "missing image" when you cached them was because you were generating them with user or usr as the target argument, so only that person got the asset.

You'd have to use either ```GLOB.clients``` or ```world``` as the user arg to icon2html.
However, that is only everyone connected at the moment, so you would need to extract the asset sending part of icon2html out, so whenever you read the cached image you always sent it out to those who don't have it (the asset cache uses browse() so it won't send the files to those who already have them)

A quick run down:
```
proc/build_path_icon(path, user)
    if(we have a cached icon for path)
        send_asset(cached icon) to user
        return cached icon
    else
        make icon with icon2html(icon, GLOB.clients)
        cache icon
        return icon
```

Reply:
Thanks for the tip, what I just did is roughly equivalent to this & seems to work in-game (and doesn't cause flicker anymore).